075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # ...
+status:
+  cruiseControl:
+    autoRebalance:
+      state: # if a scaling operation is running or what's the state of the current rebalancing
+      add-brokers: # list with brokers' IDs to add (as part of the scaling up)
+      remove-brokers: # list with brokers' IDs to remove (as part of the scaling down)
+```
+
+The `state` field provides the current state of the overall operation: if scaling is going on or what's the phase of a rebalancing operation.
+When the rebalancing is running, its value is got from the status of the corresponding `KafkaRebalance` custom resource that was created for the purpose.
+The `add-brokers` and `remove-brokers` track the brokers' IDs that are part of the scaling (up/down) actions to be useful across reconciliations, while scaling or the effective rebalancing is going on.
+Their values are got from the new `KafkaCluster.addedNodes` method (needs to be implemented) and the already existing `KafkaCluster.removedNodes` one.
+
+### Auto-rebalancing execution
+
+Before describing the process in more details, it's worth mentioning how scaling up and down happens today within the Cluster Operator reconciliation flow.
+
+The scaling up is triggered asynchronously during a reconciliation which isn't blocked on waiting for the newly added nodes to be ready.
+In the `KafkaReconcile.reconcile()` method, it is triggered via the `podSet()` method which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of spinning up the newly added nodes while the current reconciliation just ends.
+It means that, after the scaling up is triggered, the new pods will take time to be up, running and ready, so it's likely that the auto-rebalancing will happen on a future reconciliation.
+
+The scaling down runs synchronously within a single reconciliation which is blocked on waiting for the nodes to be removed.
+In the `KafkaReconcile.reconcile()` method, it is executed via the `scaleDown()` method which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of removing the nodes while the current reconciliation waits for the end of the operation.
+
+The idea is about adding a new `autoRebalancing()` method in the `KafkaReconcile.reconcile()` flow, taking care of starting a rebalancing and checking the status of an ongoing one across reconciliations.
+The `KafkaReconcile.reconcile()` flow, taking into account methods involved in the scaling and rebalancing, would be something like:
+
+* ... 
+* `scaleDown()`
+* ...
+* `podSet()`
+* ...
+* `autoRebalancing()`
+* ...
+
+> In case of scaling down, the rebalancing has to happen before removing brokers and not after that as it looks like from the above flow. Despite that, some internal state will help to have the right actions order across reconciliations as explained in the next sections.
```

Comment:
I do not understand this comment. I'm not sure what are you trying to say. Also, the `autoRebalancing()` here is clearly after the scale-down. You probably need to explain how the scale-down prevention works here and why despite being placed after the scale-down, it will not be done after the scale down. You should likely also touch whether you plan to do any validation for the scale-down prevention disablement or you leave it to the user.

Reply:
I was just highlighting that the `autoRebalancing()` method placing doesn't reflect the order of execution. I mean, it's correct for scaling up but it looks like not for scaling down. If I put it before the `scaleDown()` method, it looks ok for scaling down but not for scaling up anymore :-)
It's just an heads up about something I explain in the "Auto-rebalancing on scaling down" section, around line 223 where I said "On the **1st** reconciliation (scaling down is skipped, rebalancing started)".

Reply:
But it is not any internal state that helps with it. The scale-down prevention will revert it until the nodes are emptied by the autorebalance.

Reply:
I said it on the another comment -> but it fits here as well -> my expectation was that this would not happen synchronously as part of Kafka reconciliation. But you seem to plan that. Is that the case? I don't think that would work well because of the timing etc.